### DISCUSSION TITLE
`tar_quarto()` returns `Error : ! System command 'quarto' failed`

### DISCUSSION THREAD

**Discussion by turbanisch** [ropensci/tarchetypes#116] [Help]
I am trying to include a Quarto website project as a target but keep getting the error `Error : ! System command 'quarto' failed`. I am working in an R Studio Project (which is where the targets pipeline is defined) and the homepage files are another R Studio Project in a subfolder called `homepage`.  For now, the homepage files do not invoke any other targets. 

Edit: It turned out that the culprit was a file path I used to embed an image with 
```r
knitr::include_graphics("img/my_image.jpg")
```
Adding `homepage/` to the path (i.e. making it relative to the top-level project in which the targets pipeline lives) made it work. Perhaps a comment in the targets user manual would be useful? It would also be great if the error message could be more specific but I see that this might be hard to implement. 

For my case, it is probably best practice to define a new target for the image file rather than hard-coding it anyways.

**Comment by wlandau** [comment]
A `tar_quarto()` target calls `quarto::quarto_render()`, so it is best to test with the latter first. As for error messages, the Quarto stack currently does not forward them up, so it depends on https://github.com/quarto-dev/quarto-r/issues/16.

**Comment by turbanisch** [comment]
Thank you so much! I followed your suggestion and tried rendering the files both with `tar_make()` and `quarto::quarto_render("homepage/")`. With the latter, all file paths seem to be based on `homepage/` as root. With `tar_make()`, I have to add `homepage/` to the file paths when reading in files.

I now encounter a problem that has probably to do with the object type: I can load and embed a figure (created as a ggplot2 object) with both render commands using 
```r
readr::read_rds("homepage/import/figure.rds")
```
but loading and embedding a table (created as a gt table object) the same way only works with `quarto::quarto_render("homepage/")`. With `tar_make()` I get the error message from above. 

Given that the error seems to depend on object type, could it have to do with the packages targets is loading in the background? I do include the gt package in `tar_options_set()`. It also works in a dummy document that I knit without targets (via point-and-click in R Studio):
````
---
title: Test
output: html
---

```{r}
readr::read_rds("homepage/import/table.rds")
```
`````
It also works when I replace the gt table object with a simple dataframe.

**Reply by wlandau** [reply]
> Thank you so much! I followed your suggestion and tried rendering the files both with tar_make() and quarto::quarto_render("homepage/"). With the latter, all file paths seem to be based on homepage/ as root. With tar_make(), I have to add homepage/ to the file paths when reading in files.

The pipeline needs to each target from the project root (`execute_dir = getwd()` in `quarto_render()`) to ensure the report can load any dependent targets from the `_targets/` data store.

> I now encounter a problem that has probably to do with the object type: I can load and embed a figure (created as a ggplot2 object) with both render commands using

Some objects like `gt` tables need the package loaded to render. You could call `library(gt)` in a code chunk for that.

**Comment by turbanisch** [comment]
Thanks again! Calling `library(gt)` in the chunk doesn't seem to help though. 
I created a minimal example here: https://github.com/turbanisch/quarto-target-test The pipeline only creates a single gt-formatted table object which is then embedded in `index.qmd` in the `homepage` folder. If I run the code as it is, the table object is printed as a list. When I add `library(gt)`, I get the error `Error : ! System command 'quarto' failed` again. Adding the chunk option `#| results: asis` doesn't seem to help either (and isn't generally necessary when embedding gt objects in RMarkdown/Quarto documents). 
Sorry for taking so much of your time - hopefully this isn't just something that is not working on my machine.

**Reply by wlandau** [reply]
I looked back, and it is caused by an issue in `gt`:

* https://github.com/ropensci/targets/discussions/920#discussioncomment-3527527
* https://github.com/rstudio/gt/issues/1040
* https://github.com/rstudio/gt/pull/1063
* https://github.com/rstudio/gt/pull/1084
* https://github.com/quarto-dev/quarto-cli/issues/2996

A solution could be to upgrade the Quarto CLI or downgrade `gt` to 0.6.0.

**Comment by turbanisch** [comment]
Updating Quarto CLI solved it! Thanks for all the time you put into this, I truly appreciate it. Not sure I would have been able to figure this out by myself at all (I switched to `knitr::kable()` in the meantime).

**Comment by turbanisch** [comment]
I have another question that is likely related: When I try to embed an image into the Quarto document (via `knitr::include_graphics("homepage/img/Rplot.png")`), the pipeline throws an error if I define the image path relative to the qmd document (i.e., without "homepage/" in the path). When I define it relative to the project (where the pipeline lives) as above, the pipeline runs without error but the image path is broken:
<img width="848" alt="grafik" src="https://user-images.githubusercontent.com/51969412/202054165-39e7a402-0f14-4bb6-beba-4f8de68b1741.png">
Apparently the image path in the resulting html file is not adapted (`<p><img src="homepage/img/Rplot.png" class="img-fluid"></p>`). Do you know of a way to fix this? I can't think of another way to define the relative file path. I put the example here: https://github.com/turbanisch/quarto-target-test  
I would also be interested in a solution that tracks the file path as a target and embeds it via `tar_read()` but in my experiments the same issue occured.

**Reply by wlandau** [reply]
It's the same issue: `targets` needs to render the site with `execute_dir` at the project root, but the figures are referenced using relative paths. Regular Markdown syntax `![](./img/Rplot.png)` and `withr::with_dir("homepage", knitr::include_graphics("./img/Rplot.png"))` both seem to work. 

> I would also be interested in a solution that tracks the file path as a target and embeds it via tar_read() but in my experiments the same issue occured.

Something like `withr::with_dir("homepage", knitr::include_graphics(tar_read(image_file)))` would probably do it.

**Reply by turbanisch** [reply]
Thanks again! It works fine without `tar_read()` now but I can't get it to work with the last approach you suggested. Probably it's fine just embedding the image files outside the pipeline but I was trying to understand the mechanics of it. Since `tar_read(image_file)` already returns a file path that starts with "homepage/" I was wondering if we would not have to remove this part in order to make it relative to the homepage directory (or move up with `..`)?

**Reply by wlandau** [reply]
Yeah, either you would need to remove the `"homepage"` part or call `tar_read(image_file)` separately from `![](./img/Rplot.png)` (declaring the dependency and embedding the image in different lines).
